File: docs/litterboxed.py
#%%

from collections import defaultdict
from itertools import combinations, permutations, product
import random
from bisect import bisect_left
from datetime import date


def make_check():
    words = load_words("scrabble.txt", False)

    def check(x):
        x = x.lower()
        i = bisect_left(words, x)
        if i != len(words) and words[i] == x:
            return True
        return False

    return check


def partial(func, /, *args, **keywords):
    def newfunc(*fargs, **fkeywords):
        newkeywords = {**keywords, **fkeywords}
        return func(*args, *fargs, **newkeywords)

    newfunc.func = func
    newfunc.args = args
    newfunc.keywords = keywords
    return newfunc


def _check(w):
    if len(w) < 3:
        return False
    last = None
    for c in w:
        if c == last:
            return False
        last = c
    return True


def load_words(name, check=True):
    with open(name) as f:
        wordlist = f.read().split()
    if not check:
        return wordlist
    return list(filter(_check, wordlist))


def make_triples(words):
    ans = defaultdict(list)
    for w in words:
        if len(w) == 3 == len(set(w)):
            ans["".join(sorted(w))].append(w)
    return ans


# %%


def solve_cover(X, solution=None):
    if solution is None:
        solution = []
    if not X:
        yield list(solution)
    else:
        c = min(X, key=lambda c: len(X[c]))
        for r in list(X[c]):
            solution.append(r)
            cols = select(X, r)
            for s in solve_cover(X, solution):
                yield tuple(s)
            deselect(X, r, cols)
            solution.pop()


def select(X, r):
    cols = []
    for j in r:
        for i in X[j]:
            for k in i:
                if k != j:
                    X[k].remove(i)
        cols.append(X.pop(j))
    return cols


def deselect(X, r, cols):
    for j in reversed(r):
        X[j] = cols.pop()
        for i in X[j]:
            for k in i:
                if k != j:
                    X[k].add(i)


def gen_pair(words, wbfl, unique):
    while True:
        w1 = random.choice(words)
        w2 = random.choice(wbfl[w1[-1]])
        if len(set(w1 + w2)) == 12 and (not unique or len(w1 + w2) == 13):
            return w1, w2


def partition(w1, w2, triples=None):
    w = w1 + w2[1:]
    Y = []
    for tri in combinations(sorted(set(w)), 3):
        if any("".join(x) in w for x in permutations(tri, 2)):
            continue
        tri = "".join(tri)
        if triples is not None:
            if tri not in triples:
                continue
            tri = random.choice(triples[tri])
        Y.append(tri)
    X = defaultdict(set)
    for y in Y:
        for x in y:
            X[x].add(y)
    return next(solve_cover(X), None)


def _generate(words, wbfl, unique, triples=None):
    c = 0
    while True:
        c += 1
        w1, w2 = gen_pair(words, wbfl, unique)
        sol = partition(w1, w2, triples)
        if sol is not None:
            return sol


def make_generate(words="easy.txt", unique=True):
    random.seed(str(date.today()))

    if isinstance(words, str):
        words = load_words(words)
    # triples = make_triples(words)
    wbfl = defaultdict(list)  # words by first letter
    for w in words:
        wbfl[w[0]].append(w)
    return partial(_generate, words, wbfl, unique)


# %%


def make_word_filter(puzzle):
    letter_to_id = {}
    for i, t in enumerate(puzzle):
        for c in t:
            letter_to_id[c] = i

    def word_filter(w):
        last = None
        for c in w:
            i = letter_to_id.get(c, None)
            if i is None or i == last:
                return False
            last = i
        return True

    return word_filter


def _solve(words, puzzle, unique=False):
    ltp = defaultdict(lambda: ([], []))
    for w in filter(make_word_filter(puzzle), words):
        ltp[w[0]][1].append(w)
        ltp[w[-1]][0].append(w)
    for l1, l2 in ltp.values():
        for w1, w2 in product(l1, l2):
            if len(set(w1 + w2)) == 12 and (not unique or len(w1 + w2) == 13):
                yield w1, w2


def make_solve():
    return partial(_solve, load_words("easy.txt"))


# %%
# def good_puzzle(bound):
#     while True:
#         puzzle = generate()
#         it = solve(puzzle)
#         c = 0
#         for sol in it:
#             c += 1
#             if c > bound:
#                 break
#         else:
#             return puzzle


# while True:
#     puzzle = good_puzzle(5)
#     print(*puzzle)
#     print(*next(solve(puzzle)))
#     print("-" * 10)

File: src/litterboxed.js
function sfc32(a, b, c, d) {
    return function () {
        a >>>= 0;
        b >>>= 0;
        c >>>= 0;
        d >>>= 0;
        var t = (a + b) | 0;
        a = b ^ b >>> 9;
        b = c + (c << 3) | 0;
        c = (c << 21 | c >>> 11);
        d = d + 1 | 0;
        t = t + d | 0;
        c = c + t | 0;
        return (t >>> 0) / 4294967296;
    }
}

function xmur3(str) {
    for (var i = 0, h = 1779033703 ^ str.length; i < str.length; i++) {
        h = Math.imul(h ^ str.charCodeAt(i), 3432918353);
        h = h << 13 | h >>> 19;
    }
    return function () {
        h = Math.imul(h ^ (h >>> 16), 2246822507);
        h = Math.imul(h ^ (h >>> 13), 3266489909);
        return (h ^= h >>> 16) >>> 0;
    }
}

let rand;
export let seed = (seed) => {
    let gen = xmur3(seed);
    rand = sfc32(gen(), gen(), gen(), gen());
}


function randomChoice(arr) {
    return arr[Math.floor(arr.length * rand())];
}

export let loadWords = async (path) => fetch(path)
    .then(response => response.text()).then(t => t.split('\n'))


let solveCover = (Y) => {
    // Y is a list of list of elements
    let X = new Map();
    for (let y of Y) {
        for (let x of y) {
            if (!X.has(x))
                X.set(x, new Set());
            X.get(x).add(y);
        }
    }
    let solution = new Array();
    let rec = () => {
        if (X.size == 0)
            return solution;
        let c = X.keys().next().value;
        for (const x of X.keys()) {
            if (X.get(x).size < X.get(c).size)
                c = x;
        }
        for (let r of X.get(c)) {
            solution.push(r);
            let cols = select(X, r);
            let ans = rec();
            if (ans) return ans;
            deselect(X, r, cols)
            solution.pop();
        }
        return false;
    }
    return rec();
}

let select = (X, r) => {
    let cols = new Array();
    for (let j of r) {
        for (let i of X.get(j))
            for (let k of i)
                if (k != j)
                    X.get(k).delete(i);
        cols.push(X.get(j)), X.delete(j);
    }
    return cols;
}
let deselect = (X, r, cols) => {
    for (let j of Array.from(r).reverse()) {
        X.set(j, cols.pop());
        for (let i of X.get(j))
            for (let k of i)
                if (k != j)
                    X.get(k).add(i);
    }
}


let unorderedTriples = (set) => {
    let ans = new Array();
    let ar = Array.from(new Set(set));
    for (let i = 0; i < ar.length; i++) {
        for (let j = i + 1; j < ar.length; j++) {
            for (let k = j + 1; k < ar.length; k++) {
                ans.push(ar[i] + ar[j] + ar[k]);

            }
        }
    }
    return ans;
}


let pairs = (set) => {
    let ar = Array.from(new Set(set));
    let ans = new Array();
    for (let i = 0; i < ar.length; i++) {
        for (let j = i + 1; j < ar.length; j++) {
            ans.push(ar[i] + ar[j]);

        }
    }
    return ans
}
let partition = (w1, w2) => {
    let w = w1 + w2.slice(1);
    let Y = new Array();
    for (let tri of unorderedTriples(w)) {
        if (pairs(tri).some(x => w.indexOf(x) != -1))
            continue;
        Y.push(tri);
    }
    return solveCover(Y);
}
let genPair = (words, wbfl, unique) => {
    while (1) {
        let w1 = randomChoice(words);
        let w2 = randomChoice(wbfl.get(w1.slice(-1)))
        if (new Set(w1 + w2).size == 12 && (!unique || (w1 + w2).length == 13))
            return [w1, w2];
    }
}
export let getDate = () => {
    var today = new Date();
    return (
        today.getFullYear() +
        "-" +
        (today.getMonth() + 1) +
        "-" +
        today.getDate()
    );
};

export let makeGenerate = (words) => {
    seed(getDate());
    let wbfl = new Map();
    for (let w of words) {
        let c = w[0];
        if (!wbfl.has(c))
            wbfl.set(c, new Array());
        wbfl.get(c).push(w);
    }
    return () => {
        while (true) {
            let [w1, w2] = genPair(words, wbfl, true);
            let puzzle = partition(w1, w2);
            if (puzzle)
                return puzzle;
        }
    }

}

export let makeCheck = (words) => (w) => {
    w = w.toLowerCase();
    let a = 0,
        b = words.length;
    while (b - a > 1) {
        let c = Math.floor((a + b) / 2);
        if (w < words[c])
            b = c;
        else
            a = c;
    }
    return w == words[a];
};